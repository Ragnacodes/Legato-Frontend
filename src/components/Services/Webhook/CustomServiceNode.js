import React from 'react';
import CustomNode from '../CustomNode';
import NodeServiceIcon from '../../ServiceIcons/NodeServiceIcon';

const CustomServiceNode = (props) => {
    const shape = (
        <NodeServiceIcon service='webhooks' />
    );

    return <CustomNode shape={shape} {...props} />;
};

export default CustomServiceNode;