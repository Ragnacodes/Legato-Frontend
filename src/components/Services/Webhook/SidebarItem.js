import React from 'react';
import { Language } from '@material-ui/icons';
import CustomSidebarItem from '../CustomSidebarItem';

const SidebarItem = () => {
    return (
        <CustomSidebarItem
            icon={<Language fontSize="large" className="icon webhook" />}
            serviceName="webhook"
        />
    );
};

export default SidebarItem;