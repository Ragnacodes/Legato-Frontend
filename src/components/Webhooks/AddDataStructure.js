import React, { useState } from 'react';
import {
  TextField,
  Button,
  Card,
  CardActions,
  CardContent,
  Typography,
} from '@material-ui/core';

const AddDataStructure = ({ setVisible }) => {
  const [dsInfo, setDsInfo] = useState({
    items: [
      { key: 1, tye: 'Array', label: 'item1' },
      { key: 2, type: 'Text', label: 'item2' },
      { key: 3, type: 'Text', label: 'item2' },
    ],
  });
  const handleDelete = (chipToDelete) => {
    setDsInfo((prev) => {
      return {
        ...prev,
        items: prev.items.filter((chip) => chip.key !== chipToDelete.key),
      };
    });
  };
  const handleChangeName = (e) => {
    setDsInfo((prev) => {
      return {
        ...prev,
        name: e.target.value,
      };
    });
  };
  return (
    <div className="ds-popover">
      <Typography className="ds-popover-title" variant="h5">
        Add Data Structure
      </Typography>
      <TextField
        className="edit-wh-field"
        onChange={handleChangeName}
        name="name"
        label="Data Structure Name"
        variant="outlined"
        size="small"
        defaultValue={dsInfo['name']}
        error={'name' in dsInfo && !dsInfo['name']}
        helperText={'name' in dsInfo ? !dsInfo['name'] && 'Required.' : ''}
      />
      <Typography variant="h6" component="h2">
        Items
      </Typography>
      <Card className="ds-items-card" variant="outlined">
        <CardContent className="ds-items-card-content">
          {dsInfo.items.map((item) => {
            return (
              <Card variant="outlined" className="ds-item">
                <CardContent>
                  {Object.keys(item).map((p) => (
                    <Typography variant="body2">{item[p]}</Typography>
                  ))}
                </CardContent>
              </Card>
              // <Chip
              //   size="small"
              //   className="ds-item"
              //   label={item.label}
              //   color="primary"
              //   onDelete={() => handleDelete(item)}
              // />
            );
          })}
        </CardContent>
        <CardActions>
          <Button variant="contained" color="primary" size="small">
            Add Item
          </Button>
        </CardActions>
      </Card>
      <div className="wh-ds-actions">
        <Button autoFocus onClick={() => setVisible(false)} color="primary">
          Cancel
        </Button>
        <Button
          onClick={() => {
            // handleSave(info);
            setVisible(false);
          }}
          variant="contained"
          color="primary"
        >
          Save
        </Button>
      </div>
    </div>
  );
};

export default AddDataStructure;
