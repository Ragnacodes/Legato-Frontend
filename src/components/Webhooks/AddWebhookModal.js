import React from 'react';
import {
  Typography,
  Dialog,
  DialogContent,
  DialogTitle,
} from '@material-ui/core';
import WebhookSettings from './WebhookSettings';

const AddWebhookModal = ({ visible, handleSave, setVisible }) => {
  const webhook = {
    name: "New Webhook",
    ip_restrictions: "",
    get_request_headers: false,
    get_request_http: false,
    json_passthrough: false,
  };
  return (
    <Dialog
      disableBackdropClick
      className="edit-wh-dialog"
      open={visible}
      onClose={() => setVisible(false)}
      aria-labelledby="edit-wh-dialog"
    >
      <DialogTitle disableTypography={true} className="edit-wh-dialog-title">
        <Typography variant="h5">Create a Webhook</Typography>
      </DialogTitle>
      <DialogContent>
        <WebhookSettings
          webhook={webhook}
          visible={visible}
          handleSave={handleSave}
          setVisible={setVisible}
        />
      </DialogContent>
    </Dialog>
  );
};

export default AddWebhookModal;
