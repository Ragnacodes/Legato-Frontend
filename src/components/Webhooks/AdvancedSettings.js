import React from "react";
import {
  Checkbox,
  FormControlLabel,
  Typography,
} from "@material-ui/core";
import InfoIcon from "@material-ui/icons/Info";
// import AddDataStructure from "./AddDataStructure";
const AdvancedSettings = ({ handleChange, info }) => {
  // const dsMenu = useRef(null);
  // const addDsPopper = useRef(null);
  // const [addDsPopperOpen, setAddDsPopperOpen] = useState(false);
  // const [dsMenuOpen, setDsMenuOpen] = useState(false);

  return (
    <div className="wh-advanced-settings">
      {/* <div className="data-structure-field">
        <FormControl
          className="edit-wh-field select-ds-field"
          size="small"
          variant="outlined"
        >
          <InputLabel id="demo-simple-select-outlined-label">
            Data structure
          </InputLabel>
          <Select
            labelId="demo-simple-select-outlined-label"
            id="demo-simple-select-outlined"
            value=""
            //   onChange={handleChange}
            label="Data structure"
          >
            <MenuItem value="">
              <em>None</em>
            </MenuItem>
            <MenuItem value={10}>Ten</MenuItem>
            <MenuItem value={20}>Twenty</MenuItem>
            <MenuItem value={30}>Thirty</MenuItem>
          </Select>
        </FormControl>
        <IconButton
          aria-label="edit-data-structures"
          fontSize="small"
          size="small"
          onClick={() => setDsMenuOpen(true)}
          // onMouseDown={(e) => e.preventDefault()}
        >
          <MoreVertIcon ref={dsMenu} />
        </IconButton>
        <Menu
          id="simple-menu"
          anchorEl={dsMenu.current}
          keepMounted
          open={dsMenuOpen}
          onClose={() => setDsMenuOpen(false)}
        >
          <MenuItem onClick={() => setDsMenuOpen(false)}>Edit</MenuItem>
          <MenuItem onClick={() => setDsMenuOpen(false)}>Clone</MenuItem>
        </Menu>
        <IconButton
          size="small"
          aria-label="add-data-structures"
          className="add-icon"
          onClick={() => setAddDsPopperOpen(true)}
          // onMouseDown={(e) => e.preventDefault()}
        >
          <AddIcon ref={addDsPopper} />
        </IconButton>
        <Popover
          disableBackdropClick
          open={addDsPopperOpen}
          anchorEl={addDsPopper.current}
          onClose={() => setAddDsPopperOpen(false)}
          anchorOrigin={{
            vertical: "top",
            horizontal: "right",
          }}
          transformOrigin={{
            vertical: "top",
            horizontal: "left",
          }}
        >
          <AddDataStructure setVisible={setAddDsPopperOpen} />
        </Popover>
      </div> 
      <Typography className="help-text" variant="body2">
        <InfoIcon fontSize="small" className="help-icon" />
        Data structure to be used for validation of incoming data. Leave empty
        if you want us to just pass on the received data without any validation.
      </Typography>
      */}
      <FormControlLabel
        control={
          <Checkbox
            name="get_request_headers"
            checked={info["get_request_headers"]}
            color="primary"
            onChange={(e) => handleChange(e.target.name, e.target.checked)}
          />
        }
        label="Get request headers"
      />
      <Typography className="help-text" variant="body2">
        <InfoIcon fontSize="small" className="help-icon" />
        The headers will be added to the request body.
      </Typography>
      <FormControlLabel
        control={
          <Checkbox
            name="get_request_http"
            checked={info["get_request_http"]}
            color="primary"
            onChange={(e) => handleChange(e.target.name, e.target.checked)}
          />
        }
        label="Get request HTTP method"
      />
      <Typography className="help-text" variant="body2">
        <InfoIcon fontSize="small" className="help-icon" />
        The HTTP method (verb) will be added to the request body.
      </Typography>
      <FormControlLabel
        control={
          <Checkbox
            name="json_passthrough"
            checked={info["json_passthrough"]}
            color="primary"
            onChange={(e) => handleChange(e.target.name, e.target.checked)}
          />
        }
        label="JSON pass-through"
      />
      <Typography className="help-text" variant="body2">
        <InfoIcon fontSize="small" className="help-icon" />
        If enabled, JSON payloads are returned as a string.
      </Typography>
    </div>
  );
};

export default AdvancedSettings;
