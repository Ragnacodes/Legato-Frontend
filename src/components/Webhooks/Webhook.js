import React from "react";
import { useState } from "react";
import { connect } from "react-redux";
import {
  Button,
  ListItem,
  ListItemText,
  ListItemSecondaryAction,
  Switch,
  Link,
  TextField,
  Typography,
  InputAdornment,
  IconButton,
  Tooltip,
} from "@material-ui/core";
import SaveAltIcon from "@material-ui/icons/SaveAlt";
import DeleteIcon from "@material-ui/icons/Delete";
import ClearIcon from "@material-ui/icons/Clear";
import EditIcon from "@material-ui/icons/Edit";
import LocalShippingIcon from "@material-ui/icons/LocalShipping";
import * as actions from "../../actions/webhooks";
import EditModal from "./WebhookSettingsModal";
import WebhookQueue from "./WebhookQueue";
import header from "../../utils/api-header";
import axios from "axios";
import { errorNotification, successNotification } from "../Notification";
const Webhook = ({ webhook, updateWebhook }) => {
  const { id, name, active, url, queue, ...other } = webhook;
  const [renameToggle, setRenameToggle] = useState(false);
  const [modifiedName, setName] = useState(name);
  const [editModalVisible, setEditModalVisible] = useState(false);
  const [queueModal, setQueueModal] = useState(false);
  const onTextFieldChange = (e) => {
    setName(e.target.value);
  };
  const saveNewName = () => {
    handleUpdateWebhook({ name: modifiedName });
    // renameWebhook(id, modifiedName);
    setRenameToggle(false);
  };

  const cancelNewName = () => {
    setRenameToggle(false);
    setName(name);
  };

  const handleToggleState = () => {
    handleUpdateWebhook({ enable: !active });
  };

  // const handleRemoveWebhook = () => {
  //   removeWebhook(id);
  // };

  const handleUpdateWebhook = (data) => {
    console.log(header);
    updateWebhook(webhook.id, { ...data, active: data.enable });
    axios
      .patch(
        `http://localhost:8080/api/users/d/services/webhook/18f9453d-f1a2-40ff-8d3c-5a7f12e90b6c`,
        {
          name: data.name,
          enable: data.enable,
        },
        {
          headers: {
            Authorization:
              "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImQiLCJleHAiOjE2MTgwMjYzMjN9.EUxm2NDgX4ox2BFuJzY1LqHMBsh1haSOuLhVjKs_l5g",
          },
        }
      )
      .then((response) => {
        console.log(response);
        successNotification("Updated successfully.");
        updateWebhook(webhook.id, data);
      })
      .catch((error) => {
        if (error.response) {
          let str = error.response.data.message;
          errorNotification(
            "Unable to update: " +
              str.charAt(0).toUpperCase() +
              str.slice(1) +
              "."
          );
        } else {
          errorNotification("Unable to update: " + error.message);
        }
      });
  };

  return (
    <ListItem className="wh-item">
      <EditModal
        webhook={webhook}
        visible={editModalVisible}
        handleSave={handleUpdateWebhook}
        setVisible={setEditModalVisible}
      />
      <ListItemText
        primary={
          renameToggle ? (
            <TextField
              className="wh-name-field"
              onChange={onTextFieldChange}
              name="webhook-name"
              variant="outlined"
              size="small"
              defaultValue={name}
              error={!modifiedName}
              // helperText={!modifiedName && "Required."}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    {modifiedName && (
                      <IconButton
                        aria-label="save-name"
                        onClick={saveNewName}
                        onMouseDown={(e) => e.preventDefault()}
                        edge="end"
                      >
                        <SaveAltIcon />
                      </IconButton>
                    )}
                    <IconButton
                      aria-label="save-name"
                      onClick={cancelNewName}
                      onMouseDown={(e) => e.preventDefault()}
                      edge="end"
                    >
                      <ClearIcon />
                    </IconButton>
                  </InputAdornment>
                ),
              }}
            />
          ) : (
            <Typography
              className="wh-name"
              onClick={() => setRenameToggle(true)}
              variant="body1"
            >
              {name}
            </Typography>
          )
        }
        secondary={<Link href={url}>{url}</Link>}
        className="name"
      />

      <ListItemSecondaryAction className="wh-actions">
        <Button
          size="small"
          variant="contained"
          color="primary"
          disableElevation
          className="edit-button"
          startIcon={<EditIcon />}
          onClick={() => setEditModalVisible(true)}
        >
          Edit
        </Button>
        <Tooltip title={active ? "Disable?" : "Enable?"} placement="top">
          <Switch
            edge="end"
            onChange={handleToggleState}
            checked={active}
            color="primary"
            size="small"
            className="switch"
          />
        </Tooltip>
        <Button
          size="small"
          variant="contained"
          color="primary"
          disableElevation
          className="queue-button"
          startIcon={<LocalShippingIcon />}
          onClick={() => setQueueModal(true)}
        >
          {queue.length}
        </Button>
        <WebhookQueue
          queue={queue}
          visible={queueModal}
          setVisible={setQueueModal}
        />
        {/* <Tooltip title="Delete Webhook" placement="right">
          <IconButton
            aria-label="delete"
            className="delete-button"
            color="primary"
            onClick={handleRemoveWebhook}
          >
            <DeleteIcon fontSize="small" />
          </IconButton>
        </Tooltip> */}
      </ListItemSecondaryAction>
    </ListItem>
  );
};

const mapStateToProps = (state) => ({
  // webhook: props.webhook,
});

const mapDispatchToProps = (dispatch) => ({
  updateWebhook: (id, data) => dispatch(actions.updateWebhook(id, data)),
  // renameWebhook: (id, data) => dispatch(actions.renameWebhook(id, data)),
  // toggleWebhookState: (id) => dispatch(actions.toggleWebhookState(id)),
  // removeWebhook: (id) => dispatch(actions.removeWebhook(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Webhook);
