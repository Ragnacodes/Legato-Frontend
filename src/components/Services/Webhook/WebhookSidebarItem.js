import React from 'react';
import { Language } from '@material-ui/icons';
import SidebarItem from '../../Sketchpad/Sidebar/SidebarItem';
import trigger1 from './Trigger1/trigger1';
import action1 from './Action1/action1';

const WebhookSidebarItem = () => {
    const triggers = [trigger1];
    const actions = [action1];

    return (
        <SidebarItem
            icon={<Language fontSize="large" className="icon webhook" />}
            service="webhook"
            triggers={triggers}
            actions={actions}
        />
    );
};

export default WebhookSidebarItem;