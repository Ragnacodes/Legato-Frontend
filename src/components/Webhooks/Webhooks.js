import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import { Container, Divider, Button } from "@material-ui/core";
import List from "@material-ui/core/List";
import Webhook from "./Webhook";
import Axios from "../../utils/axiosConfig";
import { setWebhooks, addWebhook } from "../../actions/webhooks";
import Appbar from "../Layout/Appbar";
import AddWebhookModal from "./AddWebhookModal";
import { errorNotification, successNotification } from "../Layout/Notification";
import PageTitle from "../Layout/PageTitle";

const Webhooks = ({ webhooks, username, setWebhooks, addWebhook }) => {
  const [addModalVisible, setAddModalVisible] = useState(false);
  useEffect(() => {
    Axios.get(`/users/${username}/services/webhooks`)
      .then((response) => {
        setWebhooks(response.data.webhooks);
      })
      .catch((error) => {
        console.log(error);
      });
  }, [setWebhooks, username]);

  const addNewWebhook = (data) => {
    Axios.post(`/users/${username}/services/webhooks`, {
      name: data.name,
    })
      .then((response) => {
        // setWebhooks(response.data);
        console.log(response.data);
        let str = response.data.message;
        successNotification(str.charAt(0).toUpperCase() + str.slice(1) + ".");
        addWebhook(response.data.webhook);
      })
      .catch((error) => {
        console.log(error);
        if (error.response) {
          let str = error.response.data.message;
          errorNotification(
            "Unable to update: " +
              str.charAt(0).toUpperCase() +
              str.slice(1) +
              "."
          );
        } else {
          errorNotification("Unable to update: " + error.message);
        }
      });
  };

  return (
    <>
      <Appbar
        rightChildren={
          <Button
            variant="contained"
            color="secondary"
            onClick={() => setAddModalVisible(true)}
          >
            Add Webhook
          </Button>
        }
        leftChildren={<PageTitle title="Webhooks" />}
      />
      <main className="main">
        <div className="app-bar-spacer" />
        <div className="content-container">
          <AddWebhookModal
            visible={addModalVisible}
            setVisible={setAddModalVisible}
            handleSave={addNewWebhook}
          />
          <Container maxWidth="lg" className="webhooks-list">
            <List>
              {webhooks.map((w) => {
                return (
                  <div key={w.id}>
                    <Webhook webhook={w} />
                    <Divider />
                  </div>
                );
              })}
            </List>
          </Container>
        </div>
      </main>
    </>
  );
};

const mapStateToProps = (state) => ({
  webhooks: state.webhooks.webhooks,
  username: state.auth.username,
});

const mapDispatchToProps = (dispatch) => ({
  setWebhooks: (data) => dispatch(setWebhooks(data)),
  addWebhook: (data) => dispatch(addWebhook(data)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Webhooks);
