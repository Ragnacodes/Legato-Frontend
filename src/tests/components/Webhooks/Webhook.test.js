import React from 'react';
import { shallow } from 'enzyme';
import { Webhook } from '../../../components/Webhooks/Webhook';
import { ListItem } from '@material-ui/core';
let wrapper, props, webhook, updateWebhook, deleteWebhook;
beforeEach(() => {
  webhook = { name: 'name', id: 1, url: 'url', isEnable: true };
  updateWebhook = jest.fn();
  deleteWebhook = jest.fn();
  props = {
    webhook,
    updateWebhook,
    deleteWebhook,
  };
  wrapper = shallow(<Webhook {...props} />);
});

test('should render Webhook component correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render only one list item', () => {
  expect(wrapper.find(ListItem).length).toBe(1);
});
