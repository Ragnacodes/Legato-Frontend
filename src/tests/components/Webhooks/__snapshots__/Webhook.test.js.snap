// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Webhook component correctly 1`] = `
<WithStyles(ForwardRef(ListItem))
  className="wh-item"
  key="[object Object]"
>
  <WebhookSettingsModal
    handleSave={[Function]}
    setVisible={[Function]}
    visible={false}
    webhook={
      Object {
        "id": 1,
        "isEnable": true,
        "name": "name",
        "url": "url",
      }
    }
  />
  <WithStyles(ForwardRef(ListItemText))
    className="name"
    primary={
      <OnClickTextField
        defaultText="name"
        divClassName=""
        handleCancel={[Function]}
        handleSave={[Function]}
        textfieldSize="small"
      />
    }
    secondary={
      <WithStyles(ForwardRef(Tooltip))
        arrow={true}
        classes={
          Object {
            "popper": "custom-helper-tooltip",
          }
        }
        onClose={[Function]}
        placement="bottom-start"
        title="Copy to Clipboard"
      >
        <WithStyles(ForwardRef(Link))
          onClick={[Function]}
        >
          url
        </WithStyles(ForwardRef(Link))>
      </WithStyles(ForwardRef(Tooltip))>
    }
  />
  <WithStyles(ForwardRef(ListItemSecondaryAction))
    className="wh-actions"
  >
    <WithStyles(ForwardRef(Button))
      className="edit-button"
      color="primary"
      disableElevation={true}
      onClick={[Function]}
      size="small"
      startIcon={<Memo />}
      variant="contained"
    >
      Edit
    </WithStyles(ForwardRef(Button))>
    <WithStyles(ForwardRef(Tooltip))
      placement="top"
      title="Disable?"
    >
      <WithStyles(ForwardRef(Switch))
        checked={true}
        className="switch"
        color="primary"
        edge="end"
        onChange={[Function]}
        size="small"
      />
    </WithStyles(ForwardRef(Tooltip))>
    <WithStyles(ForwardRef(IconButton))
      className="queue-button"
      color="primary"
      onClick={[Function]}
      size="small"
      variant="contained"
    >
      <Memo([object Object]) />
    </WithStyles(ForwardRef(IconButton))>
    <WebhookQueue
      queue={
        Array [
          Object {
            "created_at": "Fri Apr 09 2021 22:44:31",
            "data": Object {},
            "id": 1,
            "scenarios": "12345",
            "size": "1B",
            "type": "webhook",
          },
          Object {
            "created_at": "Fri Apr 09 2021 22:44:31",
            "data": Object {
              "key": "value",
            },
            "id": 2,
            "scenarios": "12345",
            "size": "2B",
            "type": "webhook",
          },
          Object {
            "created_at": "Fri Apr 09 2021 22:44:31",
            "data": Object {
              "key": "value",
              "key2": "value2",
            },
            "id": 3,
            "scenarios": "12345",
            "size": "3B",
            "type": "webhook",
          },
        ]
      }
      setVisible={[Function]}
      visible={false}
    />
    <WithStyles(ForwardRef(Tooltip))
      placement="right"
      title="Delete Webhook"
    >
      <WithStyles(ForwardRef(IconButton))
        aria-label="delete"
        className="delete-button"
        color="primary"
        onClick={[Function]}
      >
        <Memo([object Object])
          fontSize="small"
        />
      </WithStyles(ForwardRef(IconButton))>
    </WithStyles(ForwardRef(Tooltip))>
    <YesNoModal
      handleNo={[Function]}
      handleYes={[Function]}
      setVisible={[Function]}
      text="Delete name?"
      visible={false}
    />
  </WithStyles(ForwardRef(ListItemSecondaryAction))>
</WithStyles(ForwardRef(ListItem))>
`;
